### DISCUSSION TITLE
Compiled Bindings, ItemsRepeater, and IEnumerables?

### DISCUSSION THREAD

**Discussion by jasin** [AvaloniaUI/Avalonia#11709] [Q&A] [answered]
I have a Dictionary<string,decimal> 

I can't seem to understand how to access the Key/Values inside of the DataTemplate when using Compiled Bindings. I know I can set the x:DatatType to a class etc, or Cast to a class type. I know this about the x:DataType, but how do I set it to my ViewModel property Accounts so I can access the KeyValuePairs?

I do not see that really explained in the documentation or am I missing something?
```<ItemsRepeater Items="{CompiledBinding Accounts}">
          <ItemsRepeater.ItemTemplate>
            <DataTemplate>
              <Grid ColumnDefinitions="*, Auto">
                <TextBlock Grid.Column="0" Text="{CompiledBinding AccountName}"/>
                <TextBlock Grid.Column="1" Text="{CompiledBinding AccountBalance, StringFormat={}{0:C2}}"/>
              </Grid>
            </DataTemplate>
          </ItemsRepeater.ItemTemplate>
        </ItemsRepeater>```

**Comment by timunie** [comment]
not sure if that is supported at all. But if, I'd expect it is a Key-Value-Pair, so use it like this: 

```xml
          <ItemsRepeater.ItemTemplate>
            <DataTemplate>
              <Grid ColumnDefinitions="*, Auto">
                <TextBlock Grid.Column="0" Text="{CompiledBinding Key}"/>
                <TextBlock Grid.Column="1" Text="{CompiledBinding Value, StringFormat={}{0:C2}}"/>
              </Grid>
            </DataTemplate>
          </ItemsRepeater.ItemTemplate>
        </ItemsRepeater>
```

Otherwise, you may need to write a converter for your items for example

**Reply by jasin** [reply]
If I switch out the compiled binding for a reflection binding, Key/Value work but when I use it with CompileBinding I get an exception

`System.NullReferenceException`

Can you give me a rough idea what a converter would look like? I see there really isn't anything in the documentation about converters and the one sample I found doesn't really give me a good idea of how I would need to approach this. All though, I assume since I am getting a Null Reference Exception there won't be much a value converter would accomplish?

**Reply by jasin** [reply]
I've managed to get this far. But I now get this error `x:Type extension should take exactly one constructor parameter without any content OR TypeName property`

```<ItemsRepeater Items="{CompiledBinding Accounts}">
          <ItemsRepeater.ItemTemplate>
            <DataTemplate DataType="{x:Type generic:IDictionary(x:String, x:Decimal)}">
              <Grid ColumnDefinitions="*, Auto">
                <TextBlock Grid.Column="0" Text="{CompiledBinding Key}"/>
                <TextBlock Grid.Column="1" Text="{CompiledBinding Value, StringFormat={}{0:C2}}"/>
              </Grid>
            </DataTemplate>
          </ItemsRepeater.ItemTemplate>
        </ItemsRepeater>```

**Reply by timunie** [reply]
Can you try out this datatype? https://learn.microsoft.com/en-us/dotnet/api/system.collections.dictionaryentry?view=net-7.0

**Reply by jasin** [reply]
Yep, that works, thank you (See Below)

**Reply by jasin** [reply]
I take that back, as soon as switched them back to CompiledBinding I get a InvalidCastException from type KeyValuePair to DictionaryEntry.

**Comment by timunie** [comment, marked as answer]
@jasin I think I found a working solution while looking around in older issues: https://github.com/AvaloniaUI/Avalonia/issues/391#issuecomment-979819573 

in the ViewModel I have: 

```cs
public Dictionary<string, string> Test { get; } = new()
{
    {"Hello", "World"},
    {"Abc", "123"},
    {"Def", "456"},
};
``` 

And in the View I have to specify DataType as following
```xml
<ListBox ItemsSource="{ReflectionBinding Test}">
    <ListBox.ItemTemplate>
        <DataTemplate>
            <!-- This is the important part. It doesn't work inline -->
            <DataTemplate.DataType>
                <x:Type TypeName="generic:KeyValuePair" x:TypeArguments="x:String, x:String"/>
            </DataTemplate.DataType>

            <StackPanel>
                <TextBlock Text="{Binding Key}"></TextBlock>
                <TextBlock Text="{Binding Value, Mode=OneWay}"></TextBlock>
            </StackPanel>
        </DataTemplate>
    </ListBox.ItemTemplate>
</ListBox>
```

Result:
![image](https://github.com/AvaloniaUI/Avalonia/assets/47110241/1847eef0-715e-49d9-99b1-0d47e95f033c)

**Reply by jasin** [reply]
This in fact does work. 

At least I don't feel completely dumb, I was on the right track, but doing outside of inlining was the trick. Thank you!
I feel like this should be documented somewhere.
